akka-projection-grpc/src/main/scala/akka/projection/grpc/internal/EventConsumerServiceImpl.scala

```diff
@@ -0,0 +1,94 @@
+/*
+ * Copyright (C) 2009-2023 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.projection.grpc.internal
+
+import akka.NotUsed
+import akka.actor.typed.ActorRef
+import akka.actor.typed.ActorSystem
+import akka.actor.typed.scaladsl.LoggerOps
+import akka.pattern.StatusReply
+import akka.persistence.EventWriter
+import akka.projection.grpc.internal.proto.ConsumerEvent
+import akka.projection.grpc.internal.proto.ConsumerEventAck
+import akka.projection.grpc.internal.proto.EventConsumerService
+import akka.stream.scaladsl.Source
+import akka.stream.typed.scaladsl.ActorFlow
+import akka.util.ByteString
+import akka.util.Timeout
+import org.slf4j.LoggerFactory
+
+import java.net.URLEncoder
+import scala.concurrent.duration.DurationInt
+
+/**
+ * INTERNAL API
+ *
+ * gRPC push protocol service for the consuming side
+ */
+private[akka] object EventConsumerServiceImpl {
+
+  def directJournalConsumer(journalPluginId: String, persistenceIdTransformer: String => String)(
+      implicit system: ActorSystem[_]): EventConsumerServiceImpl = {
+    // FIXME is this name unique, could we create multiple for the same journal? (we wouldn't be able to bind them to the same port)
+    val eventWriter = system.systemActorOf(
+      EventWriter(journalPluginId),
+      s"EventWriter-${URLEncoder.encode(journalPluginId, ByteString.UTF_8)}")
+
+    new EventConsumerServiceImpl(eventWriter, persistenceIdTransformer)
+  }
+
+}
+
+/**
+ * INTERNAL API
+ */
+private[akka] final class EventConsumerServiceImpl(
+    eventWriter: ActorRef[EventWriter.Command],
+    persistenceIdTransformer: String => String)(implicit system: ActorSystem[_])
+    extends EventConsumerService {
+
+  private val logger = LoggerFactory.getLogger(classOf[EventConsumerServiceImpl])
+
+  private val protoAnySerialization = new ProtoAnySerialization(system)
+
+  private implicit val timeout: Timeout = 5.seconds // FIXME from config or can we get rid of it
+
+  override def consumeEvent(in: Source[ConsumerEvent, NotUsed]): Source[ConsumerEventAck, NotUsed] = {
+    // FIXME do we need to make sure events for the same pid are ordered? Should be single writer per pid but?
+    in.prefixAndTail(1)
+      .flatMapConcat {
+        case (Seq(ConsumerEvent(ConsumerEvent.Message.Init(init), _)), tail) =>
+          logger.info("Event stream from [{}] started", init.originId)
+          tail.collect {
+            case ConsumerEvent(ConsumerEvent.Message.Event(event), _) => event
+          }
+        case (_, _) =>
+          throw new IllegalArgumentException(
+            "Expected stream in starts with Init event followed by events but got something else")
+      }
+      .via(ActorFlow.askWithStatus(8)(eventWriter) {
+        (in: proto.Event, replyTo: ActorRef[StatusReply[EventWriter.WriteAck]]) =>
+          // FIXME would we want request metadata/producer ip/entire envelope in to persistence id transformer?
```

Comment:
yes, one case could be that based on the origin of the event you want to adjust the persistenceId to include that origin (for example making it more unique)

Reply:
we could have a transformation function of the Entire EventEnvelope, but we shouldn't allow changing the seqNr

btw, would then probably be nice with some "copy methods" on the EventEnvelope, such as `withTags`, `withPersistenceId`.

Reply:
I guess we could accept a transform but always check that sequence number was not changed. Or more a more limited transform API instead of adding API to let users actually transform the entire envelope, `Transform.addTags()`.

Reply:
Added a try at a transform API but didn't include the origin as a parameter yet: fdb350ed667754ce5a817b744da3b4d2baa049fc

Reply:
And now the transform can be created per origin/event stream based on origin id and request metadata [fda1176](https://github.com/akka/akka-projection/pull/932/commits/fda1176910daebed09cd84b0796c88785a88d95d)